File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include<bits/stdc++.h>
using namespace std;
const int dy[] = {-1,0,1,0};
const int dx[] = {0,1,0,-1};
int tc,m,n,k,ret,ny,nx;
int a[54][54];
int visited[54][54];
void DFS(int sy,int sx)
{
	visited[sy][sx]=1;
	for(int i=0;i<4;i++)
	{
		ny = sy + dy[i];
		nx = sx + dx[i];
		if(ny<0 || ny>=n || nx<0 || nx>=m) continue;
		if(a[ny][nx]==1 && visited[ny][nx]==0)
		{
			DFS(ny,nx);
		}
	}
	return;
}
int main()
{
	cin>>tc;
	for(int i=0;i<tc;i++)
	{
		memset(a,0,sizeof(a));
		memset(visited,0,sizeof(visited));
		cin>>m>>n>>k;
		for(int i=0;i<k;i++)
		{
			int y,x;
			cin>>x>>y;
			a[y][x]=1;
		}
		ret=0;
		for(int i=0;i<n;i++)
		{
			for(int j=0;j<m;j++)
			{
				if(a[i][j]==1 && !(visited[i][j]))
				{
					ret++;
					DFS(i,j);
				}
			}
		}
		cout<<ret<<"\n";
	}
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
sys.setrecursionlimit(10000)

dy=[-1,0,1,0]
dx=[0,1,0,-1]

def dfs(y, x):
    visited[y][x]=1
    for i in range(4):
        ny = y+dy[i]
        nx = x+dx[i]
        if ny<0 or ny>=n or nx<0 or nx>=m or visited[ny][nx] or a[ny][nx]==0:
            continue
        if not visited[ny][nx] and a[ny][nx]==1:
            dfs(ny,nx)

t=int(input().strip())

for _ in range(t):
    ret=0
    n, m, k = map(int, input().split())
    a = [[0] * m for _ in range(n)]
    visited = [[0] * m for _ in range(n)]

    for _ in range(k):
        y,x=map(int,input().split())
        a[y][x]=1

    for i in range(n):
        for j in range(m):
            if not visited[i][j] and a[i][j]==1:
                ret+=1
                dfs(i,j)

    print(ret)